### ISSUE TITLE
Rule Change: [no-loop-func] add an option to ignore IIFEs

### ISSUE THREAD

**Issue by bradzacher** [eslint/eslint#16902]
### What rule do you want to change?

no-loop-func

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
/* eslint no-loop-func: 'error' */
let current = getStart();
while (current) {
  (() => {
    current;
    current.a;
    current.b;
    current.c;
    current.d;
  })();
  
  current = current.upper;
}
```

### What does the rule currently do for this code?

the rule will report for this code even though this code is provably safe because the function is immediately invoked.

### What will the rule do after it's changed?

the rule will not report for this code as it is provably safe

### Participation

- [X] I am willing to submit a pull request to implement this change.

### Additional comments

[Playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLWxvb3AtZnVuYzogJ2Vycm9yJyAqL1xubGV0IGN1cnJlbnQgPSBnZXRTdGFydCgpO1xud2hpbGUgKGN1cnJlbnQpIHtcbiAgKCgpID0+IHtcbiAgICBjdXJyZW50O1xuICAgIGN1cnJlbnQuYTtcbiAgICBjdXJyZW50LmI7XG4gICAgY3VycmVudC5jO1xuICAgIGN1cnJlbnQuZDtcbiAgfSkoKTtcbiAgXG4gIGN1cnJlbnQgPSBjdXJyZW50LnVwcGVyO1xufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnsiZXM2Ijp0cnVlfX19)

Also worth noting that the rule's error message is wonky here. For the example code it reports the same ID many times:
```
Function declared in a loop contains unsafe references to variable(s) 'current', 'current', 'current', 'current', 'current'. 
```

**Comment by mdjermanovic** [comment]
I support the proposal to ignore IIFEs, and I think there's no need for an option (the issue title suggests adding an option).

However, named function expressions could be stored so perhaps we should additionally check if the function is referenced.

```js
let current = getStart();
const arr = [];
while (current) {
  (function f() {
    current;
    arr.push(f); // <--
  })();
  
  current = current.upper;
}
```

**Comment by fasttime** [comment]
> I support the proposal to ignore IIFEs, and I think there's no need for an option (the issue title suggests adding an option).
> 
> However, named function expressions could be stored so perhaps we should additionally check if the function is referenced.

I think we should also keep reporting on IIFEs with async and generator functions, because their asynchronous state may become accessible from outside, even without a self-reference.

```js
let current = getStart();
const arr = [];
while (current) {
  const p = (async () => {
    await someDelay();
    current;
  })();

  arr.push(p);
  current = current.upper;
}
await Promise.all(arr);
```

Other that this (and possibly other yet-to-be-discovered edge cases), I also like the proposal.

**Comment by sam3k** [comment]
In the 2023-03-23 TSC meeting we have agreed to: 

1. Accept the proposal
2. Not add an option and instead make this change a default behavior 
3. Continue warning in the async/generator case

Notes from this meeting will be published in this [PR](https://github.com/eslint/tsc-meetings/pull/419/files).

**Comment by Gautam-Arora24** [comment]
Would love to work on this issue

**Comment by Gautam-Arora24** [comment]
Apparently, we should also ignore reporting for this rule, if the function is called (irrespective of its IIFE nature) anywhere inside the loop's scope.
